### PR TITLE
Disable default marshalling for bool

### PR DESCRIPTION
**Pull request by jkoritzinsky**
For consistency with the new default rules for `bool` fields in value types, update the default marshalling for `bool` parameters and return values to be 1-byte non-normalized instead of 4-byte normalized.

I've tried to split up the commits to make them easier to review:

1. updates all existing source-generated P/Invokes to have explicit `[MarshalAs(UnmanagedType.Bool)]` attributes to preserve the current default behavior.
2. updates the default marshalling rules in the generator.
3. updates the conversion code-fix.
4. updates the compatibility doc.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
I agree with this change, but I think a discussion with other stakeholders is warranted due to the scope of impact.

**Comment by jkotas** [comment]
1-byte default for bool is likely wrong more than 50% of the time and it will have to be overridden most of the time. Given that, I think it would be better to make the marshaling directive for bool mandatory.

It matches the approach that we took with string marshaling. There is no great default for string marshaling and so we require it to be always specified explicitly.

**Comment by jkotas** [comment]
We may want to introduce a new `BOOL` type (exact name and shape subject to negotiation - for example, it can be `enum BOOL` that we use internally in some places) to make the typical bool marshaling patterns less verbose and more efficient.

**Comment by elinor-fung** [comment]
@jkoritzinsky had also suggested exposing a `BOOL` type  - similar to our internal enum - last Friday. Definitely agree it would be a good idea (just don't think any of us have gotten to actually proposing it).

**Comment by tannergooding** [comment]
> We may want to introduce a new BOOL type (exact name and shape subject to negotiation - for example, it can be enum BOOL that we use internally in some places) to make the typical bool marshaling patterns less verbose and more efficient.

This is another scenario where, in my opinion, having some `[TransparentStructAttribute]` would be beneficial. Having an `enum BOOL` is ok but it isn't extensible, you can't ever expose additional operators or methods, and it's only one of many types where a regular `wrapper struct` isn't possible. Users will have to find/discover it and the name will conflict with similar types exposed in other libraries already.

You have cases like `BOOL`, `COLORREF`, `LPARAM, `WPARAM`, and all the `HANDLE` types on Win32. You have types like `VkBool32` and `VkBuffer` for Vulkan as an xplat API. You have similar examples in various Linux or MacOS centric APIs.

There are plenty of common cases, and a near endless number of other scenarios, where in C users have some `typedef primitive Name;` and where for clarity, readability, or even to expose functionality that only makes sense for `Name` and not `primitive` in general you want to provide some kind of wrapper in C# and to have it ultimately be treated as blittable.

`TransparentStructAttribute` (or a better name) would allow the same scenario as it does in Rust (`repr(transparent)`) and would allow users to define their own wrappers that fit the needs of their libraries and for the needs of their customers. It would allow the current special-casing of `CLong`, `CULong`, and `NFloat` to be generalized around such an attribute instead and for users to fill the gap on their own in the future.

**Comment by jkotas** [comment]
Transparent structures can be implemented with a bit of boilerplate code already: https://github.com/dotnet/runtime/blob/main/docs/design/libraries/DllImportGenerator/StructMarshalling.md#special-case-transparent-structures. We can make the source generator generate this boilerplate for you if it moves the needle enough.

**Comment by jkoritzinsky** [comment]
I've updated this pr to disable default marshalling for bool instead of changing the default.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshalAsMarshallingGeneratorFactory.cs [hunk 1]
```diff
@@ -75,8 +75,6 @@ public MarshalAsMarshallingGeneratorFactory(InteropGenerationOptions options, IM
                 case { ManagedType: PointerTypeInfo(_, _, IsFunctionPointer: true), MarshallingAttributeInfo: NoMarshallingInfo or MarshalAsInfo(UnmanagedType.FunctionPtr, _) }:
                     return s_blittable;
 
-                case { ManagedType: SpecialTypeInfo { SpecialType: SpecialType.System_Boolean }, MarshallingAttributeInfo: NoMarshallingInfo }:
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshalAsMarshallingGeneratorFactory.cs:78]: Can we make the diagnostic have a detailed message stating that specifying marshalling info is required for `bool` (similar to what we do for `char`/`string`)?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingAttributeInfo.cs [hunk 2]
```diff
@@ -780,6 +780,17 @@ private CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashS
                 return true;
             }
 
+            if (type.SpecialType == SpecialType.System_Boolean)
+            {
+                // We explicitly don't support marshalling bool or without any marshalling info
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingAttributeInfo.cs:785]: This comment seems muddled.

File: src/libraries/System.Runtime.InteropServices/tests/DllImportGenerator.Tests/BooleanTests.cs [hunk 3]
```diff
@@ -19,67 +20,67 @@ partial class NativeExportsNE
...
+        [GeneratedDllImport(NativeExportsNE_Binary, EntryPoint = "winbool_return_as_uint")]
         public static partial uint ReturnUIntBoolAsUInt([MarshalAs(UnmanagedType.U4)] bool input);
 
-        [GeneratedDllImport(NativeExportsNE_Binary, EntryPoint = "bool_return_as_uint")]
+        [GeneratedDllImport(NativeExportsNE_Binary, EntryPoint = "winbool_return_as_uint")]
         public static partial uint ReturnWinBoolAsUInt([MarshalAs(UnmanagedType.Bool)] bool input);
 
-        [GeneratedDllImport(NativeExportsNE_Binary, EntryPoint = "bool_return_as_uint")]
+        [GeneratedDllImport(NativeExportsNE_Binary, EntryPoint = "bytebool_return_as_uint")]
         public static partial uint ReturnDefaultBoolAsUInt(bool input);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/DllImportGenerator.Tests/BooleanTests.cs:33]: I assume this should be removed now?

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by elinor-fung** [comment]
Generator changes look good to me. We should also run libraries outerloop (once the innerloop is green).

**Comment by jkoritzinsky** [comment]
Failures are https://github.com/dotnet/runtime/issues/66100

**Pull request merged**